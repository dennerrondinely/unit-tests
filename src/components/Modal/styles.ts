import styled, { css, DefaultTheme, keyframes } from 'styled-components';
import media from 'styled-media-query';
import { ModalProps } from '.';

type ContainerProps = Pick<ModalProps, 'size' | 'open'>;
type ContentProps = Pick<ModalProps, 'size'>;

const slideUp = keyframes`
  {
    0% {
        transform: translateY(100%);
    }
    100% {
        transform: translateY(0px);
    }
  }
`;

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    min-height: calc(${theme.spacings.xxlarge} * 4);
    width: ${theme.sizes.small};
  `,
  medium: (theme: DefaultTheme) => css`
    min-height: calc(${theme.spacings.xxlarge} * 5);
    width: ${theme.sizes.medium};
  `,
  large: (theme: DefaultTheme) => css`
    min-height: calc(${theme.spacings.xxlarge} * 7);
    width: ${theme.sizes.large};
  `
};

export const Wrapper = styled.div<ContainerProps>`
  position: fixed;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  justify-content: center;
  align-items: center;

  ${({ theme, open }) => css`
    display: ${open ? 'flex' : 'none'};
    z-index: ${theme.layers.modal};
  `}
`;

export const Background = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 1;
  opacity: 0.8;
  ${({ theme }) => css`
    background-color: ${theme.colors['gray-dark']};
  `}
`;

export const Title = styled.div`
  ${({ theme }) => css`
    margin-bottom: ${theme.spacings.small};
  `}
`;

export const Content = styled.div<ContentProps>`
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  z-index: 2;
  ${({ theme, size = 'medium' }) => css`
    font-weight: ${theme.font.normal};
    color: ${theme.colors['gray-dark']};
    background-color: ${theme.colors.white};
    padding: ${theme.spacings.medium} ${theme.spacings.medium}
      ${theme.spacings.small};
    font-size: ${theme.font.sizes.small};
    line-height: 1.8rem;
    width: 100%;
    height: 100%;
    animation: ${slideUp} 0.3s ease-in-out;

    ${media.greaterThan('large')`
      animation: none;
      height: auto;
      border-radius: ${theme.border.radius.medium};
      ${wrapperModifiers[size](theme)};
    `}
  `}
`;

export const ButtonWrapper = styled.div`
  display: inline-flex;
  justify-content: space-between;
  margin: auto 0 0;
  gap: ${({ theme }) => theme.spacings.small};

  ${media.greaterThan('large')`
    justify-content: flex-end;
  `}
`;

export const IconWrapper = styled.button`
  display: inline;
  position: absolute;
  border: none;
  cursor: pointer;
  ${({ theme }) => css`
    background-color: ${theme.colors.transparent};
    top: ${theme.spacings.small};
    right: ${theme.spacings.small};
  `}
`;
