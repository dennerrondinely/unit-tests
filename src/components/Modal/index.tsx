import React, { ReactNode } from 'react';
import Button from 'components/Button';
import { Close } from 'styled-icons/material-outlined';
import Typography from 'components/Typography';

import { SizeType } from 'types/sizes';

import * as S from './styles';

export type ModalProps = {
  size?: SizeType;
  open?: boolean;
  cancel?: string;
  approve?: string;
  title?: string;
  children?: ReactNode;
  onClose?(): void;
  onApprove?(): void;
  onCancel?(): void;
};

const Modal = ({
  children,
  open,
  size = 'medium',
  cancel,
  approve,
  title,
  onClose,
  onApprove,
  onCancel
}: ModalProps) => {
  const handleClose = () => !!onClose && onClose();
  const handleCancel = () => !!onCancel && onCancel();
  const handleApprove = () => !!onApprove && onApprove();

  return (
    <S.Wrapper open={open}>
      <S.Background onClick={handleClose} />
      <S.Content size={size}>
        {title && (
          <S.Title>
            <Typography variant="subtitle" color="gray-dark" align="start">
              {title}
            </Typography>
          </S.Title>
        )}
        <S.IconWrapper onClick={handleClose}>
          <Close color="gray" size={16} />
        </S.IconWrapper>
        {children}
        <S.ButtonWrapper>
          {!!cancel && (
            <Button bg="transparent" color="gray" onClick={handleCancel}>
              {cancel}
            </Button>
          )}
          {!!approve && <Button onClick={handleApprove}>{approve}</Button>}
        </S.ButtonWrapper>
      </S.Content>
    </S.Wrapper>
  );
};

export default Modal;
