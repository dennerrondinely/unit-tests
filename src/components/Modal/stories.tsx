import React, { useState } from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';

import Modal, { ModalProps } from './index';
import Button from 'components/Button';

export default {
  title: 'Modal',
  component: Modal,
  args: {
    title: 'Why do we use it?'
  }
} as Meta;

export const Default: Story<ModalProps> = (args) => (
  <Modal {...args}>Why do we use it?</Modal>
);

export const WithButtons: Story<ModalProps> = (args) => {
  const [open, setOpen] = useState(false);
  const handleModal = () => setOpen(!open);
  return (
    <div>
      <Button onClick={handleModal}>Open Modal</Button>
      <Modal {...args} open={open} onClose={handleModal}>
        Why do we use it?
      </Modal>
    </div>
  );
};

WithButtons.args = {
  cancel: 'Cancelar',
  approve: 'Aprovar'
};
