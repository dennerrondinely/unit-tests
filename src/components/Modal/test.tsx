import React from 'react';

import { render, fireEvent, screen } from 'utils/test-utils';

import Modal from './index';

describe('<Modal />', () => {
  it('should render the modal open true', () => {
    const { container } = render(<Modal open>Modal</Modal>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the modal and click event in close button', () => {
    const onClose = jest.fn();
    render(
      <Modal open onClose={onClose}>
        Modal
      </Modal>
    );

    fireEvent.click(screen.getByRole('button'));
    expect(onClose).toHaveBeenCalled();
  });

  it('should render the modal with button cancel', () => {
    const { container } = render(
      <Modal open cancel="cancelar">
        Modal
      </Modal>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the modal with button cancel and click event', () => {
    const onCancel = jest.fn();
    render(
      <Modal open cancel="cancelar" onCancel={onCancel}>
        Modal
      </Modal>
    );

    fireEvent.click(screen.getByText('cancelar'));
    expect(onCancel).toHaveBeenCalled();
  });

  it('should render the modal with button approve', () => {
    const { container } = render(
      <Modal open approve="aprovar">
        Modal
      </Modal>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the modal with button approve and click event', () => {
    const onApprove = jest.fn();
    render(
      <Modal open approve="aprovar" onApprove={onApprove}>
        Modal
      </Modal>
    );

    fireEvent.click(screen.getByText('aprovar'));
    expect(onApprove).toHaveBeenCalled();
  });
});
