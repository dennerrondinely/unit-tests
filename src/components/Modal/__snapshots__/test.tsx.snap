// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should render the modal open true 1`] = `
.c4 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: fixed;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  z-index: 40;
}

.c1 {
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 1;
  opacity: 0.8;
  background-color: #575453;
}

.c2 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  z-index: 2;
  font-weight: 400;
  color: #575453;
  background-color: #ffffff;
  padding: 2rem 2rem 1.6rem;
  font-size: 1.6rem;
  line-height: 1.8rem;
  width: 100%;
  height: 100%;
  -webkit-animation: bOuwtO 0.3s ease-in-out;
  animation: bOuwtO 0.3s ease-in-out;
}

.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: auto 0 0;
  gap: 1.6rem;
}

.c3 {
  display: inline;
  position: absolute;
  border: none;
  cursor: pointer;
  background-color: #0000;
  top: 1.6rem;
  right: 1.6rem;
}

@media (min-width:1170px) {
  .c2 {
    -webkit-animation: none;
    animation: none;
    height: auto;
    border-radius: 0.6rem;
    min-height: calc(3.2rem * 5);
    width: 55.7rem;
  }
}

@media (min-width:1170px) {
  .c5 {
    -webkit-box-pack: end;
    -webkit-justify-content: flex-end;
    -ms-flex-pack: end;
    justify-content: flex-end;
  }
}

<div
  class="c0"
  open=""
>
  <div
    class="c1"
  />
  <div
    class="c2"
  >
    <button
      class="c3"
    >
      <svg
        aria-hidden="true"
        class="c4"
        color="gray"
        fill="currentColor"
        focusable="false"
        height="16"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
        />
      </svg>
    </button>
    Modal
    <div
      class="c5"
    />
  </div>
</div>
`;

exports[`<Modal /> should render the modal with button approve 1`] = `
.c6 {
  cursor: pointer;
  white-space: nowrap;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  border-radius: 0.8rem;
  color: #ffffff;
  background-color: #2E50D4;
  font-weight: 700;
  border: 0.2rem solid #0000;
  height: 4rem;
  min-width: 7.5rem;
  font-size: 1.4rem;
  line-height: 1.8rem;
  padding: 0 1.2rem;
}

.c6:disabled {
  cursor: not-allowed;
  color: #DAD2D0;
  background-color: #F4F2F1;
}

.c4 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: fixed;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  z-index: 40;
}

.c1 {
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 1;
  opacity: 0.8;
  background-color: #575453;
}

.c2 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  z-index: 2;
  font-weight: 400;
  color: #575453;
  background-color: #ffffff;
  padding: 2rem 2rem 1.6rem;
  font-size: 1.6rem;
  line-height: 1.8rem;
  width: 100%;
  height: 100%;
  -webkit-animation: bOuwtO 0.3s ease-in-out;
  animation: bOuwtO 0.3s ease-in-out;
}

.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: auto 0 0;
  gap: 1.6rem;
}

.c3 {
  display: inline;
  position: absolute;
  border: none;
  cursor: pointer;
  background-color: #0000;
  top: 1.6rem;
  right: 1.6rem;
}

@media (min-width:1170px) {
  .c2 {
    -webkit-animation: none;
    animation: none;
    height: auto;
    border-radius: 0.6rem;
    min-height: calc(3.2rem * 5);
    width: 55.7rem;
  }
}

@media (min-width:1170px) {
  .c5 {
    -webkit-box-pack: end;
    -webkit-justify-content: flex-end;
    -ms-flex-pack: end;
    justify-content: flex-end;
  }
}

<div
  class="c0"
  open=""
>
  <div
    class="c1"
  />
  <div
    class="c2"
  >
    <button
      class="c3"
    >
      <svg
        aria-hidden="true"
        class="c4"
        color="gray"
        fill="currentColor"
        focusable="false"
        height="16"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
        />
      </svg>
    </button>
    Modal
    <div
      class="c5"
    >
      <button
        class="c6"
        color="white"
      >
        aprovar
      </button>
    </div>
  </div>
</div>
`;

exports[`<Modal /> should render the modal with button cancel 1`] = `
.c6 {
  cursor: pointer;
  white-space: nowrap;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  border-radius: 0.8rem;
  color: #DAD2D0;
  background-color: #0000;
  font-weight: 700;
  border: 0.2rem solid #0000;
  height: 4rem;
  min-width: 7.5rem;
  font-size: 1.4rem;
  line-height: 1.8rem;
  padding: 0 1.2rem;
}

.c6:disabled {
  cursor: not-allowed;
  color: #DAD2D0;
  background-color: #F4F2F1;
}

.c4 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: fixed;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  z-index: 40;
}

.c1 {
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 1;
  opacity: 0.8;
  background-color: #575453;
}

.c2 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  z-index: 2;
  font-weight: 400;
  color: #575453;
  background-color: #ffffff;
  padding: 2rem 2rem 1.6rem;
  font-size: 1.6rem;
  line-height: 1.8rem;
  width: 100%;
  height: 100%;
  -webkit-animation: bOuwtO 0.3s ease-in-out;
  animation: bOuwtO 0.3s ease-in-out;
}

.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: auto 0 0;
  gap: 1.6rem;
}

.c3 {
  display: inline;
  position: absolute;
  border: none;
  cursor: pointer;
  background-color: #0000;
  top: 1.6rem;
  right: 1.6rem;
}

@media (min-width:1170px) {
  .c2 {
    -webkit-animation: none;
    animation: none;
    height: auto;
    border-radius: 0.6rem;
    min-height: calc(3.2rem * 5);
    width: 55.7rem;
  }
}

@media (min-width:1170px) {
  .c5 {
    -webkit-box-pack: end;
    -webkit-justify-content: flex-end;
    -ms-flex-pack: end;
    justify-content: flex-end;
  }
}

<div
  class="c0"
  open=""
>
  <div
    class="c1"
  />
  <div
    class="c2"
  >
    <button
      class="c3"
    >
      <svg
        aria-hidden="true"
        class="c4"
        color="gray"
        fill="currentColor"
        focusable="false"
        height="16"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
        />
      </svg>
    </button>
    Modal
    <div
      class="c5"
    >
      <button
        class="c6"
        color="gray"
      >
        cancelar
      </button>
    </div>
  </div>
</div>
`;
